src/mpfao/3D/tdympfao3D_ts.c

```diff
@@ -373,6 +389,200 @@ PetscErrorCode TDyMPFAOIJacobian_Vertices_3DMesh(Vec Ul, Mat A, void *ctx) {
+
+    if (vertices->num_boundary_faces[ivertex] > 0 && vertices->num_internal_cells[ivertex] < 2) continue;
+
+    PetscInt vOffsetCell    = vertices->internal_cell_offset[ivertex];
+    PetscInt vOffsetFace    = vertices->face_offset[ivertex];
+
+    PetscInt npitf_bc = vertices->num_boundary_faces[ivertex];
+    PetscInt nflux_in = vertices->num_faces[ivertex] - vertices->num_boundary_faces[ivertex];
+
+    // Compute T*P
+    PetscScalar TtimesP[nflux_in + npitf_bc];
+    for (PetscInt irow=0; irow < nflux_in + npitf_bc; irow++) {
+
+      PetscInt face_id = vertices->face_ids[vOffsetFace + irow];
+      PetscInt subface_id = vertices->subface_ids[vOffsetFace + irow];
+      PetscInt num_subfaces = 4;
+
+      if (!faces->is_local[face_id]) continue;
+
+      TtimesP[irow] = TtimesP_vec_ptr[face_id*num_subfaces + subface_id];
+      if (fabs(TtimesP[irow])<PETSC_MACHINE_EPSILON) TtimesP[irow] = 0.0;
+    }
+
+    //
+    // fluxm_ij = rho_ij * (kr/mu)_{ij,upwind} * [T] * [P]  + rho_ij * (kr/mu)_{ij,upwind} * rho_ij * G
+    //
+    // For k = i or j, the jacobian is given as:
+    // d(fluxm_ij)/dP_k =
+    //                   d(rho_ij)/dP_k +   (kr/mu)_{ij,upwind}       * [T] * [P] +
+    //                     rho_ij       + d((kr/mu)_{ij,upwind})/dP_k * [T] * [P] +
+    //                     rho_ij       +   (kr/mu)_{ij,upwind}       * T_k       +
+    //                  2 * rho_ij * d(rho_ij)/dP_k *   (kr/mu)_{ij,upwind}       * G  +
+    //                      rho_ij *   rho_ij       * d((kr/mu)_{ij,upwind})/dP_k * G
+    //
+    // For k /= i or j, the jacobian is given as:
+    // d(fluxm_ij)/dP_k =
+    //                     rho_ij       +   (kr/mu)_{ij,upwind}       * T_k
+    //
+    for (PetscInt irow=0; irow<nflux_in + npitf_bc; irow++) {
+
+      PetscInt face_id = vertices->face_ids[vOffsetFace + irow];
+      PetscInt subface_id = vertices->subface_ids[vOffsetFace + irow];
+      PetscInt fOffsetCell = faces->cell_offset[face_id];
+
+      PetscInt cell_id_up = faces->cell_ids[fOffsetCell + 0];
+      PetscInt cell_id_dn = faces->cell_ids[fOffsetCell + 1];
+
+      // If using neumann bc (which is currently no-flux), then skip the face
+      if ( tdy->mpfao_bc_type == MPFAO_NEUMANN_BC  && (cell_id_up<0 || cell_id_dn <0))  continue;
+
+      PetscReal dukvr_dPup = 0.0;
+      PetscReal dukvr_dPdn = 0.0;
+      PetscReal ukvr;
+
+      PetscReal den = 0.0;
+      if (cell_id_up>=0) den += tdy->rho[cell_id_up];
+      else               den += tdy->rho_BND[-cell_id_up-1];
+      if (cell_id_dn>=0) den += tdy->rho[cell_id_dn];
+      else               den += tdy->rho_BND[-cell_id_dn-1];
+      den *= 0.5;
+
+      PetscInt num_subfaces = 4;
+      if (TtimesP[irow] + den*GravDis_ptr[face_id*num_subfaces + subface_id]< 0.0) { // up ---> dn
+        // Flow: up --> dn
+        if (cell_id_up>=0) {
+          // "up" is an internal cell
+          ukvr       = cc->Kr[cell_id_up]/tdy->vis[cell_id_up];
```

Comment:
This could be cleaned up with some temporary variables. A lot of these are pulled from arrays repeatedly. The compiler knows how to optimize this kind of stuff, but symbolic names might help the humans.